File: smallpict-js/src/helper/smallpict.ts
import fs from 'fs';
import path from 'path';
import axios from 'axios';
import { v4 as uuidv4 } from 'uuid';
import { SmallPictClient, type ImageFormat } from '@smallpict/sdk';
import { config } from '../config/config.js';
import { pool } from '../database/db.js';

export interface ImageRecord {
  id: string;
  filename: string;
  url: string;
  size: number;
  size_origin: number;
  mime_type: string;
  created_at: string;
}

export async function processUpload(
  fileBuffer?: Buffer,
  base64Input?: string,
  originalFilename?: string,
  inputMimeType?: string
): Promise<ImageRecord> {
  let data: Buffer;
  let filename = originalFilename || '';
  let mimeType = inputMimeType || 'image/jpeg';

  if (base64Input && base64Input.trim()) {
    let b64 = base64Input.trim();
    if (b64.includes(';base64,')) {
      const parts = b64.split(';base64,');
      if (parts[0].startsWith('data:')) {
        mimeType = parts[0].substring(5);
      }
      b64 = parts[1];
    } else if (b64.startsWith('data:') && b64.includes(',')) {
      const parts = b64.split(',');
      mimeType = parts[0].substring(5).split(';')[0];
      b64 = parts[1];
    }

    b64 = b64.replace(/\s+/g, '');
    data = Buffer.from(b64, 'base64');
    if (!filename) {
      filename = `image${mimeToExtension(mimeType)}`;
    }
  } else if (fileBuffer && fileBuffer.length > 0) {
    data = fileBuffer;
    if (!filename) {
      filename = 'image.jpg';
    }
  } else {
    throw new Error('No image content provided in request');
  }

  if (data.length === 0) {
    throw new Error('Image data is empty');
  }

  const extName = path.extname(filename) || mimeToExtension(mimeType);
  let baseName = path.basename(filename, extName);
  if (!baseName) baseName = 'image';

  const epochSec = Math.floor(Date.now() / 1000);
  let timestampedFilename = `${baseName}_${epochSec}${extName}`;
  const imageId = uuidv4();
  const sizeOrigin = data.length;
  let size = sizeOrigin;
  let imageUrl = '';

  const mode = config.smallpict.mode;
  const hasCredentials = !!(config.smallpict.apiKey && config.smallpict.secretKey);

  if (hasCredentials) {
    const client = new SmallPictClient({
      apiKey: config.smallpict.apiKey,
      secretKey: config.smallpict.secretKey,
      baseUrl: config.smallpict.baseUrl,
    });

    const result = await client.optimize(data, {
      filename: timestampedFilename,
      mimeType,
      format: config.smallpict.format as ImageFormat,
      quality: config.smallpict.quality,
    });

    if (result.uploadUrl) {
      const putResponse = await axios.put(result.uploadUrl, data, {
        headers: {
          'Content-Type': mimeType,
        },
        validateStatus: () => true,
        timeout: 60000,
      });

      if (putResponse.status >= 300) {
        throw new Error(`S3 upload failed with status ${putResponse.status}`);
      }
    }

    let remoteUrl = result.url;
    const jobId = result.jobId;
    const status = (result.status || '').toLowerCase();

    if (jobId && (['pending', 'processing', 'queued'].includes(status) || !remoteUrl)) {
      for (let i = 0; i < 20; i++) {
        await new Promise((res) => setTimeout(res, 2000));
        try {
          const jobStatus = await client.getJobStatus(jobId);
          const st = (jobStatus.status || '').toLowerCase();
          if (['completed', 'succeeded', 'success', 'ready', 'done'].includes(st)) {
            if (jobStatus.url) {
              remoteUrl = jobStatus.url;
            }
            if (jobStatus.bytesSaved && jobStatus.bytesSaved > 0) {
              size = sizeOrigin - Number(jobStatus.bytesSaved);
            }
            break;
          }
        } catch {
          // ignore polling errors
        }
      }
    }

    if (!remoteUrl) {
      throw new Error('Failed to get optimized image URL from SmallPict');
    }

    if (mode === 'locale' || mode === 'local') {
      const cleanUrl = remoteUrl.includes('?') ? remoteUrl.substring(0, remoteUrl.indexOf('?')) : remoteUrl;
      const remoteExt = path.extname(cleanUrl).toLowerCase();
      if (remoteExt === '.webp') {
        timestampedFilename = `${baseName}_${epochSec}.webp`;
        mimeType = 'image/webp';
      } else if (remoteExt === '.avif') {
        timestampedFilename = `${baseName}_${epochSec}.avif`;
        mimeType = 'image/avif';
      } else if (remoteExt === '.png') {
        timestampedFilename = `${baseName}_${epochSec}.png`;
        mimeType = 'image/png';
      } else if (remoteExt === '.jpg' || remoteExt === '.jpeg') {
        timestampedFilename = `${baseName}_${epochSec}.jpg`;
        mimeType = 'image/jpeg';
      }

      const uploadDir = path.resolve(config.uploadPath);
      if (!fs.existsSync(uploadDir)) {
        fs.mkdirSync(uploadDir, { recursive: true });
      }

      const dlRes = await axios.get(remoteUrl, {
        responseType: 'arraybuffer',
        timeout: 60000,
      });

      if (dlRes.status >= 300) {
        throw new Error(`Failed to download compressed image from SmallPict: HTTP ${dlRes.status}`);
      }

      const destPath = path.join(uploadDir, timestampedFilename);
      fs.writeFileSync(destPath, Buffer.from(dlRes.data));

      imageUrl = `/uploads/${timestampedFilename}`;
      size = fs.statSync(destPath).size;
    } else {
      imageUrl = remoteUrl;
    }
  } else {
    const uploadDir = path.resolve(config.uploadPath);
    if (!fs.existsSync(uploadDir)) {
      fs.mkdirSync(uploadDir, { recursive: true });
    }

    const destPath = path.join(uploadDir, timestampedFilename);
    fs.writeFileSync(destPath, data);
    imageUrl = `/uploads/${timestampedFilename}`;
  }

  const record: ImageRecord = {
    id: imageId,
    filename: timestampedFilename,
    url: imageUrl,
    size: Number(size),
    size_origin: Number(sizeOrigin),
    mime_type: mimeType,
    created_at: new Date().toISOString(),
  };

  try {
    await pool.query(
      `INSERT INTO public.images (id, filename, url, size, size_origin, mime_type, created_at)
       VALUES ($1, $2, $3, $4, $5, $6, $7)`,
      [record.id, record.filename, record.url, record.size, record.size_origin, record.mime_type, record.created_at]
    );
  } catch (err: any) {
    console.warn('Warning: Failed to save image metadata to database:', err.message);
  }

  return record;
}

function mimeToExtension(mimeType: string): string {
  switch (mimeType.toLowerCase()) {
    case 'image/png':
      return '.png';
    case 'image/webp':
      return '.webp';
    case 'image/avif':
      return '.avif';
    case 'image/gif':
      return '.gif';
    case 'image/svg+xml':
      return '.svg';
    default:
      return '.jpg';
  }
}

File: smallpict-go/helper/smallpict.go
package helper

import (
	"bytes"
	"context"
	"encoding/base64"
	"errors"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"smallpict-go/config"
	"smallpict-go/database"
	"smallpict-go/model"

	"github.com/google/uuid"
	"github.com/tuxnoob/smallpict-go"
)

func ProcessUpload(fileHeader *multipart.FileHeader, base64Input string, customFilename string) (*model.Image, error) {
	var data []byte
	var filename string
	var mimeType string

	if base64Input != "" {
		b64Content := strings.TrimSpace(base64Input)
		if idx := strings.Index(b64Content, ";base64,"); idx != -1 {
			meta := b64Content[:idx]
			if strings.HasPrefix(meta, "data:") {
				mimeType = strings.TrimPrefix(meta, "data:")
			}
			b64Content = b64Content[idx+8:]
		} else if idx := strings.Index(b64Content, ","); idx != -1 && strings.HasPrefix(b64Content, "data:") {
			mimeType = strings.TrimPrefix(b64Content[:idx], "data:")
			b64Content = b64Content[idx+1:]
		}

		b64Content = strings.ReplaceAll(b64Content, " ", "")
		b64Content = strings.ReplaceAll(b64Content, "\n", "")
		b64Content = strings.ReplaceAll(b64Content, "\r", "")

		decoded, err := base64.StdEncoding.DecodeString(b64Content)
		if err != nil {
			decoded, err = base64.RawStdEncoding.DecodeString(b64Content)
		}
		if err != nil {
			decoded, err = base64.URLEncoding.DecodeString(b64Content)
		}
		if err != nil {
			return nil, fmt.Errorf("invalid base64 image data: %w", err)
		}
		data = decoded
		filename = customFilename
	} else if fileHeader != nil {
		f, err := fileHeader.Open()
		if err != nil {
			return nil, fmt.Errorf("failed to open uploaded file: %w", err)
		}
		defer f.Close()

		data, err = io.ReadAll(f)
		if err != nil {
			return nil, fmt.Errorf("failed to read uploaded file: %w", err)
		}

		filename = customFilename
		if filename == "" {
			filename = fileHeader.Filename
		}
		if mimeType == "" {
			mimeType = fileHeader.Header.Get("Content-Type")
		}
	} else {
		return nil, errors.New("no image content provided")
	}

	if len(data) == 0 {
		return nil, errors.New("image data is empty")
	}

	if mimeType == "" || mimeType == "application/octet-stream" {
		mimeType = http.DetectContentType(data)
	}

	ext := filepath.Ext(filename)
	base := strings.TrimSuffix(filename, ext)
	if base == "" {
		base = "image"
	}
	if ext == "" {
		ext = mimeToExtension(mimeType)
	}

	epochSec := time.Now().Unix()
	timestampedFilename := fmt.Sprintf("%s_%d%s", base, epochSec, ext)
	imageID := uuid.New().String()
	sizeOrigin := int64(len(data))
	size := sizeOrigin
	var imageURL string

	cfg := config.Cfg
	mode := strings.ToLower(cfg.SmallPictMode)
	if mode == "" {
		mode = "locale"
	}

	hasCredentials := cfg.SmallPictAPIKey != "" && cfg.SmallPictSecretKey != ""

	if hasCredentials {
		client, err := smallpict.NewClient(
			smallpict.WithAPIKey(cfg.SmallPictAPIKey),
			smallpict.WithSecretKey(cfg.SmallPictSecretKey),
			smallpict.WithBaseURL(cfg.SmallPictBaseURL),
		)
		if err != nil {
			return nil, fmt.Errorf("failed to initialize smallpict client: %w", err)
		}

		ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
		defer cancel()

		opts := &smallpict.OptimizeOptions{
			Filename: timestampedFilename,
			MIMEType: mimeType,
			Format:   cfg.SmallPictFormat,
			Quality:  cfg.SmallPictQuality,
		}

		result, err := client.Optimize(ctx, bytes.NewReader(data), opts)
		if err != nil {
			return nil, fmt.Errorf("smallpict optimize failed: %w", err)
		}

		if result.UploadURL != "" {
			req, err := http.NewRequestWithContext(ctx, http.MethodPut, result.UploadURL, bytes.NewReader(data))
			if err != nil {
				return nil, fmt.Errorf("failed to create s3 upload request: %w", err)
			}
			req.Header.Set("Content-Type", mimeType)

			resp, err := http.DefaultClient.Do(req)
			if err != nil {
				return nil, fmt.Errorf("failed to upload image to s3 presigned url: %w", err)
			}
			defer resp.Body.Close()

			if resp.StatusCode < 200 || resp.StatusCode >= 300 {
				return nil, fmt.Errorf("s3 upload returned status code %d", resp.StatusCode)
			}
		}

		remoteURL := result.URL
		if result.CompressedSize > 0 {
			size = result.CompressedSize
		}

		if result.JobID != "" && (result.Status == "pending" || result.Status == "processing" || remoteURL == "") {
			for i := 1; i <= 20; i++ {
				select {
				case <-ctx.Done():
					return nil, ctx.Err()
				case <-time.After(2 * time.Second):
				}

				status, err := client.GetJobStatus(ctx, result.JobID)
				if err != nil {
					continue
				}

				st := strings.ToLower(status.Status)
				if st == "succeeded" || st == "completed" || st == "success" || st == "ready" || st == "done" {
					if status.URL != "" {
						remoteURL = status.URL
					}
					if status.BytesSaved != nil && *status.BytesSaved > 0 {
						size = sizeOrigin - *status.BytesSaved
					}
					break
				}
			}
		}

		if remoteURL == "" {
			return nil, errors.New("failed to get optimized image URL from smallpict")
		}

		if mode == "locale" || mode == "local" {
			cleanURL := remoteURL
			if idx := strings.Index(cleanURL, "?"); idx != -1 {
				cleanURL = cleanURL[:idx]
			}
			remoteExt := strings.ToLower(filepath.Ext(cleanURL))
			if remoteExt == ".webp" {
				timestampedFilename = fmt.Sprintf("%s_%d.webp", base, epochSec)
				mimeType = "image/webp"
			} else if remoteExt == ".avif" {
				timestampedFilename = fmt.Sprintf("%s_%d.avif", base, epochSec)
				mimeType = "image/avif"
			} else if remoteExt == ".png" {
				timestampedFilename = fmt.Sprintf("%s_%d.png", base, epochSec)
				mimeType = "image/png"
			} else if remoteExt == ".jpg" || remoteExt == ".jpeg" {
				timestampedFilename = fmt.Sprintf("%s_%d.jpg", base, epochSec)
				mimeType = "image/jpeg"
			}

			uploadDir := cfg.UploadPath
			if uploadDir == "" {
				uploadDir = "./uploads"
			}
			_ = os.MkdirAll(uploadDir, 0755)

			req, err := http.NewRequestWithContext(ctx, http.MethodGet, remoteURL, nil)
			if err != nil {
				return nil, fmt.Errorf("failed to create download request: %w", err)
			}

			resp, err := http.DefaultClient.Do(req)
			if err != nil {
				return nil, fmt.Errorf("failed to download compressed image from smallpict: %w", err)
			}
			defer resp.Body.Close()

			if resp.StatusCode < 200 || resp.StatusCode >= 300 {
				return nil, fmt.Errorf("download compressed image failed with status %d", resp.StatusCode)
			}

			optimizedData, err := io.ReadAll(resp.Body)
			if err != nil {
				return nil, fmt.Errorf("failed to read downloaded compressed image: %w", err)
			}

			destPath := filepath.Join(uploadDir, timestampedFilename)
			if err := os.WriteFile(destPath, optimizedData, 0644); err != nil {
				return nil, fmt.Errorf("failed to save local file: %w", err)
			}

			imageURL = fmt.Sprintf("/uploads/%s", timestampedFilename)
			size = int64(len(optimizedData))
		} else {
			imageURL = remoteURL
		}
	} else {
		uploadDir := cfg.UploadPath
		if uploadDir == "" {
			uploadDir = "./uploads"
		}
		_ = os.MkdirAll(uploadDir, 0755)

		destPath := filepath.Join(uploadDir, timestampedFilename)
		if err := os.WriteFile(destPath, data, 0644); err != nil {
			return nil, fmt.Errorf("failed to save local file: %w", err)
		}

		imageURL = fmt.Sprintf("/uploads/%s", timestampedFilename)
	}

	record := model.Image{
		ID:         imageID,
		Filename:   &timestampedFilename,
		URL:        &imageURL,
		Size:       &size,
		SizeOrigin: &sizeOrigin,
		MIMEType:   &mimeType,
		CreatedAt:  time.Now(),
	}

	if database.DB != nil {
		if err := database.DB.Create(&record).Error; err != nil {
			return nil, fmt.Errorf("failed to save image metadata to database: %w", err)
		}
	}

	return &record, nil
}

func mimeToExtension(mimeType string) string {
	switch strings.ToLower(mimeType) {
	case "image/jpeg", "image/jpg":
		return ".jpg"
	case "image/png":
		return ".png"
	case "image/webp":
		return ".webp"
	case "image/avif":
		return ".avif"
	case "image/gif":
		return ".gif"
	case "image/svg+xml":
		return ".svg"
	default:
		return ".jpg"
	}
}
